### PR TITLE
feat(experiments): Toolbar to create no-code A/B experiments

### PR DESCRIPTION
**Pull request by Phanatic**
## Problem

This PR introduces a new toolbar control to allow users to create no-code web experiments using the PostHog toolbar on their web page.  It uses the API introduced in #24872 to manage web experiments from the toolbar. 

## Changes

1. Extend the `Toolbar Elements` overlay to support selecting elements for web experiments. 
2. Add a `experiments` option to the toolbar, that is shown only when the `web-experiments` feature flag is enabled.
3. Allow users to edit existing experiments, by changing targeted HTML elements and adding/removing elements to be modified per experiment variant. 
4. Allow users to create new web experiments, directly from the toolbar.

### 1. List Web Experiments
image

### 2. Edit an existing web experiment.
image

5. Change target element for an experiment variant. 

image

## Does this work well for both Cloud and self-hosted?

Yes.

## How did you test this code?

1. Manually, in local development
2. Add storybook elements so its easy to see the visual changes.
- [x] Adding tests specifically for the ExperimentsEditingToolbarMenu.tsx now

### CODE REVIEW

File: frontend/src/toolbar/__mocks__/list-web-experiments-response.ts [hunk 1]
```diff
@@ -0,0 +1,74 @@
...
+                {
+                    transforms: [
+                        {
+                            html: '',
+                            text: 'Ola Amigo',
+                            selector: '#page > #body > .body h1',
+                            className: 'O',
+                        },
+                    ],
+                    variant_name: 'variant #0',
```
[the remark this answers is not in the document]
Comment (Phanatic) [reply frontend/src/toolbar/__mocks__/list-web-experiments-response.ts]: Fixed now!

File: frontend/src/toolbar/experiments/ExperimentsEditingToolbarMenu.tsx [hunk 2]
```diff
@@ -0,0 +1,208 @@
...
+export const ExperimentsEditingToolbarMenu = (): JSX.Element => {
+    const { selectedExperimentId, inspectingElement, experimentForm } = useValues(experimentsTabLogic)
+    const {
+        setExperimentFormValue,
+        selectExperiment,
+        selectVariant,
+        inspectForElementWithIndex,
+        deleteExperiment,
+        newVariant,
+        newElement,
```
Comment (jurajmajerik) [inline frontend/src/toolbar/experiments/ExperimentsEditingToolbarMenu.tsx:21]: I'd prefer more descriptive names for these actions: `addNewVariant`, `addNewElement`

File: frontend/src/toolbar/experiments/ExperimentsEditingToolbarMenu.tsx [hunk 3]
```diff
@@ -0,0 +1,208 @@
...
+                                <h4>{experimentForm.name}</h4>
+                            )}
+                            <LemonButton
+                                type="secondary"
+                                size="small"
+                                className="ml-3"
+                                sideIcon={<IconPlus />}
+                                onClick={newVariant}
+                            >
+                                Add Variant
```
Comment (jurajmajerik) [inline frontend/src/toolbar/experiments/ExperimentsEditingToolbarMenu.tsx:62]: Sentence case please: https://posthog.com/handbook/growth/marketing/posthog-style-guide#use-sentence-case-for-titles

```suggestion
                                Add variant
```

File: frontend/src/toolbar/experiments/ExperimentsToolbarMenu.tsx [hunk 4]
```diff
@@ -0,0 +1,72 @@
...
+import { ToolbarMenu } from '~/toolbar/bar/ToolbarMenu'
+import { ExperimentsEditingToolbarMenu } from '~/toolbar/experiments/ExperimentsEditingToolbarMenu'
+import { ExperimentsListView } from '~/toolbar/experiments/ExperimentsListView'
+import { experimentsLogic } from '~/toolbar/experiments/experimentsLogic'
+import { experimentsTabLogic } from '~/toolbar/experiments/experimentsTabLogic'
+import { toolbarConfigLogic } from '~/toolbar/toolbarConfigLogic'
+
+const ExperimentsListToolbarMenu = (): JSX.Element => {
+    const { searchTerm } = useValues(experimentsLogic)
+    // const { setSearchTerm, getActions } = useActions(actionsLogic)
```
Comment (jurajmajerik) [inline frontend/src/toolbar/experiments/ExperimentsToolbarMenu.tsx:20]:
```suggestion

```

**Comment by jurajmajerik** [comment]
Whenever possible, validate form errors in the frontend (in this case a missing selector), and use input highlighting to indicate what the error is.

If an error needs to returned from the API, display an error toast, otherwise the user doesn't know what's happening.

**Comment by jurajmajerik** [comment]
Re variants:
- the first variant must be `control` and it must be non-editable
- the second variant should already be pre-added. I'd recommend hardcoding its name to `test`

Basically we should mimic the Product Experiment form when it comes to this.

image

**Comment by jurajmajerik** [comment]
I would not prefill the experiment name as the user will always want to edit it. Use a placeholder instead.

image

**Comment by jurajmajerik** [comment]
Align vertically to the middle of container please :)

image

**Comment by jurajmajerik** [comment]
Overall, I'd like to see more visual hierarchy to the whole layout:

1. The `Remove` button removes the variant, but is right next to `Add element` - it makes me think, does it remove the element or the variant? Move it close to the variant. Then, you can even remove the `Remove` text, because it will be obvious what exactly it does (check the Product Experiment form).
2. There is no clear boundary between the individual variant sections - very little margin. Consider how to make them more distinct.
3. The `Select element` is a distinct action, so it should be formatted as a button. Right now it looks more like a search input (especially because of the magnifying glass icon).
4. `No element selected` is not an actual value, it's a placeholder. So format it differently (perhaps italics + muted color).
5. I quite like the rollout percentage information here, but add a bit more style to it (maybe a tag), otherwise it looks a bit bland.

image

**Comment by jurajmajerik** [comment]
Once you select an element, should you be able to select the same element again? I can see how this could be an issue as you'd be overwriting your existing modifications.

Is there an easy way to filter out "selectable" elements in the toolbar?

image

**Comment by jurajmajerik** [comment]
Here I've only changed the text, but the link color changed to orange. Is this expected? (might have something to do with how our own styles are targeted)

image

**Comment by jurajmajerik** [comment]
When I modify an element and then remove the element from the variant, the modification stays in place.

https://github.com/user-attachments/assets/85a4c93a-9cdc-4b64-938c-d7b16059b62b

**Comment by jurajmajerik** [comment]
For an existing experiment, I would hide this action in a `...` button in the top bar - I don't expect this to be performed _that_ often.

image

**Comment by jurajmajerik** [comment]
After saving the experiment and editing it again, the order of `control` and `test` is reversed? 🤔 

image

**Comment by jurajmajerik** [comment]
When saving 2nd experiment, I get this error:

image

It seems that feature flag key is not being passed at all - there's also no input field for this. I assume the first experiment was created with a `None` flag key, and the second experiment is trying to do the same.

**Comment by jurajmajerik** [comment]
Before saving, we should also validate that there are at least 2 variants.

image

**Comment by jurajmajerik** [comment]
When viewing an existing experiment, the transforms are not being applied to elements.

image

Similarly, in editing mode, the current state of the element only reflects the latest transform. There's no way to preview the changes for transforms that were set up earlier:

image

I wonder what the flow should be here. As a user, I want to be always able to preview any transforms for any variant.

It seems to me that we need a way to say _"Right now, I'm editing variant A, and these are all the transforms applied for it"_.

Perhaps, every variant could be an item in a collapsible accordion. When you click on a variant, that item uncollapses, and you can
- edit the transforms for that variant
- see all transforms applied _for that variant only_.

**Comment by jurajmajerik** [comment]
IMO no need to use numbering here. Users often do some sort of numbering themselves. I would sort this list by the `created_at` though (ignore if already done). Also consider adding this column to the list view, along with `created_by` and `status`.

image

**Comment by jurajmajerik** [comment]
Some comments, but this looks promising! I think it was a great idea to do this from a toolbar. I can see this being very powerful in the future 💪

**Comment by jurajmajerik** [comment]
image

The visualization is working nicely - however, I don't know which variant is being applied to the page (unless I've _just_ clicked on the button). I would at least like to see this button highlighted when the given variant is being visualized.

Have you considered structuring the variants here into collapsible accordion as suggested above? I think this would be really nice because:
- whenever a given item (variant) of the accordion is clicked & uncollapsed, the visualization would be immediately applied to the page
- the accordion would solve the problem of poor visual distinction between variants - you always know which variant you are editing, because only one is open at a time

This is what I mean, illustrated on the accordion within Surveys:

https://github.com/user-attachments/assets/6e552043-5e7b-471f-b228-3e11255753ac

Notice that at no point there's any confusion around which _question_ I'm editing - only one is open at a time, and it's also immediately visualized on the right. I think this could work great here too.

**Comment by jurajmajerik** [comment]
image

The input fields is too narrow here.

Actually, do we even need the Cancel/Save buttons here? This is just a simple text input field, why not just render it as it is?

**Comment by jurajmajerik** [comment]
image

This is vertically misaligned

**Comment by jurajmajerik** [comment]
image

Add a bit more padding to this tag. Also, I would say "50% rollout" to make it clear what it is. The text should be some lighter color.

**Comment by jurajmajerik** [comment]
https://github.com/user-attachments/assets/02d952d8-2933-4a83-a775-2d54d411c795

Clicking on Save here removes the variant.

**Comment by jurajmajerik** [comment]
A bug:
- Select the first element
- Select the second element
- Go back to the first element and try to change it to something else - it remains the same

https://github.com/user-attachments/assets/754f02c4-a8b2-46a9-9758-1123401209fc

**Review by jurajmajerik — APPROVED** [review]

**Pull request merged**